py/redrock/dataobj.py

```diff
@@ -118,8 +124,160 @@ def sharedmem_pack(self):
+        # duplicate some data from the resolution matrix.
+
+        self.targets = self._get_targets()
+
+
+    def __del__(self):
+        self.close()
+    
+    def __enter__(self):
+        return self
+    
+    def __exit__(self, type, value, tb):
+        self.close()
+        return False
+
+    def close(self):
+        return
+
+
+    def _fill_target_dictionary(self, targets):
+        n = 0
+        dictionary = OrderedDict() # keep targets as ordered in list
+        for target in targets :
+            dictionary[target.id] = {}
+            for spectra_label, spectra in zip(["spectra", "coadd"], 
+                [target.spectra, target.coadd]):
+                spectra_list = list()
+                for spectrum in spectra :
+                    spectrum_dict = {}
+                    for label, asize in zip(["wave","flux","ivar","rdata",
+                        "roffsets","rshape"], [spectrum.wave.size, 
+                        spectrum.flux.size, spectrum.ivar.size, 
+                        spectrum.R.data.size, spectrum.R.offsets.size,
+                        len(spectrum.R.shape)]):
+                        spectrum_dict[label] = [n, n + asize]
+                        n += asize
+                    spectra_list.append(spectrum_dict)
+                dictionary[target.id][spectra_label] = spectra_list
+        return dictionary, n
+
+
+    def _get_targets(self):
+        targets = list()
+        for id, tdict in self._target_dictionary.items():
+            spectra = list()
+            coadd = list()
+            for spectra_label, spectra_ref in zip(["spectra", "coadd"],
+                [spectra, coadd]):
+                spectra_dict_list = tdict[spectra_label]
+                for spectrum_dict in spectra_dict_list :
+                    # we are not doing a copy here (except maybe for R ...)
+                    wave = self._shared[spectrum_dict["wave"][0]:spectrum_dict["wave"][1]]
+                    flux = self._shared[spectrum_dict["flux"][0]:spectrum_dict["flux"][1]]
+                    ivar = self._shared[spectrum_dict["ivar"][0]:spectrum_dict["ivar"][1]]
+                    rdata = self._shared[spectrum_dict["rdata"][0]:spectrum_dict["rdata"][1]]
+                    roffsets = self._shared[spectrum_dict["roffsets"][0]:spectrum_dict["roffsets"][1]]
+                    rshape = tuple(self._shared[spectrum_dict["rshape"][0]:spectrum_dict["rshape"][1]].astype(int))
+                    rdata = rdata.reshape((roffsets.size,rdata.shape[0]//roffsets.size))
+                    spectra_ref.append( SimpleSpectrum(wave, flux, ivar, 
+                        scipy.sparse.dia_matrix((rdata, roffsets), 
+                        shape=rshape)) )            
+            targets.append(Target(id, spectra, coadd=coadd))
+
+        return targets
+
+
 class Target(object):
-    def __init__(self, targetid, spectra):
+    def __init__(self, targetid, spectra, coadd=None):
```

Comment:
Please document new `coadd` option in the docstring.  I haven't groked why this is necessary.